core-rust/core-api-server/src/core_api/helpers.rs

```diff
@@ -62,3 +64,37 @@ pub(crate) fn wrong_substate_type(substate_offset: SubstateOffset) -> ResponseEr
     }
     .into()
 }
+
+struct ByteCountWriter<'a> {
+    bytes: &'a mut usize,
+}
+
+impl<'a> ByteCountWriter<'a> {
+    fn new(bytes: &'a mut usize) -> Self {
+        Self { bytes }
+    }
+}
+
+impl<'a> Write for ByteCountWriter<'a> {
+    fn write(&mut self, data: &[u8]) -> Result<usize, std::io::Error> {
+        *self.bytes += data.len();
+        Ok(data.len())
+    }
+
+    fn flush(&mut self) -> Result<(), std::io::Error> {
+        Ok(())
+    }
+}
+
+pub trait GetJsonSize: Serialize {
+    fn get_json_size(&self) -> usize {
+        let mut bytes = 0;
+        {
+            let writer = ByteCountWriter::new(&mut bytes);
+            serde_json::to_writer(writer, &self).expect("Failed to serialize JSON");
```

Comment:
Am I right to assume that we serialize to JSON just to get final string length? That seems super wasteful. I guess we should try serialize some sort of stream that could be backed by some array/collection of transactions that would additionally account for serialized response size as collection elements get serialized.

Reply:
Right. This is what I've already spent way too much time trying to do right and efficient. Unfortunately Rust's `serde` does not support any kind of higher level stream besides the "write stream" where I lose any context of what object I'm writing.

`serde` is just plainly not really performant or meant to be used for things like this (it's interface is `-> Result<String>`), normally when this is a problem, people start using protobufs and gRPC.

Strictly for performance (but really not that accurate) not even `get_size::GetSize` can be used because of `models::SborData` dependency on `serde_json::value` itself, which can't be decorated anymore.

On the bright side though, the json serializer doesn't need to allocate memory for Strings.

Reply:
Quick question: maybe we shouldn't even bother counting bytes of JSON-serialized payload? Can we estimate TX size prior to serialization? Think:

```
val maxSize = 1 * 1024 * 1024;
var calculatedSize = 0;
val txCollection = ...;
val finalTxCollection = new List();

foreach (var tx in txCollection) {
  val txRoughSize = 1024 // some typical overhead 
    + tx.manifest.size() // user tx manifest size 
    + tx.createdSubstates.Sum(s => s.sbor_payload.size() * 1.5) // 1.5 is some constant factor
    + tx.updatedSubstates.Sum(s => s.sbor_payload.size() * 1.25) 
    + tx.events.Sum(e => e.sbor_data.size() * 1.75)

  calculatedSize += txRoughSize;
  finalTxcollection.append(tx);

  if (calculatedSize > maxSize) {
    _logger.warning("bla bla bla, max est. size reached....")

    break;
  }
}
```

Looks less fragile...

Reply:
Something like this I tried with the `get_size::GetSize`, get the RAM size and multiply by some factors. But I am stopped by `SborData::data_json` which is already an `serde_json::value` and I can't really touch (in order to add `get_size::GetSize` trait).

For sake of robustness I can still multiply current size value by some factor (just to 100% make sure that the limit is indeed an upper bound).

Reply:
Some good thoughts @KrzysztofLabus-RDX, but to keep-it-simple, I suggested this approach as a good first-stab.

In reality, the overhead of doing this is likely small in the grand scheme of things, and won't be on any critical paths if we're on an API thread here. It's also more accurate than a memory-measuring approach, particularly if we embed programmatic SBOR with user-provided names.